ductvh/ex6.ts

```diff
@@ -0,0 +1,89 @@
+/*
+
+Intro:
+
+    Filtering requirements have grown. We need to be
+    able to filter any kind of Persons.
+
+Exercise:
+
+    Fix typing for the filterPersons so that it can filter users
+    and return User[] when personType='user' and return Admin[]
+    when personType='admin'. Also filterPersons should accept
+    partial User/Admin type according to the personType.
+    `criteria` argument should behave according to the
+    `personType` argument value. `type` field is not allowed in
+    the `criteria` field.
+
+Higher difficulty bonus exercise:
+
+    Implement a function `getObjectKeys()` which returns more
+    convenient result for any argument given, so that you don't
+    need to cast it.
+
+    let criteriaKeys = Object.keys(criteria) as (keyof User)[];
+    -->
+    let criteriaKeys = getObjectKeys(criteria);
+
+*/
+
+interface User {
+  type: 'user';
+  name: string;
+  age: number;
+  occupation: string;
+}
+
+interface Admin {
+  type: 'admin';
+  name: string;
+  age: number;
+  role: string;
+}
+
+export type Person = User | Admin;
+
+export const persons: Person[] = [
+  { type: 'user', name: 'Max Mustermann', age: 25, occupation: 'Chimney sweep' },
+  { type: 'admin', name: 'Jane Doe', age: 32, role: 'Administrator' },
+  { type: 'user', name: 'Kate Müller', age: 23, occupation: 'Astronaut' },
+  { type: 'admin', name: 'Bruce Willis', age: 64, role: 'World saver' },
+  { type: 'user', name: 'Wilson', age: 23, occupation: 'Ball' },
+  { type: 'admin', name: 'Agent Smith', age: 23, role: 'Anti-virus engineer' }
+];
+
+export function logPerson(person: Person) {
+  console.log(
+      ` - ${person.name}, ${person.age}, ${person.type === 'admin' ? person.role : person.occupation}`
+  );
+}
+
+
+export function filterPersons(person: Person[], personType: User['type'], criteria: Partial<Omit<User, 'type'>>): User[]
+export function filterPersons(person: Person[], personType: Admin['type'], criteria: Partial<Omit<Admin, 'type'>>): Admin[]
+
+export function filterPersons(persons: Person[], personType: Person['type'], criteria: Partial<Omit<Person, 'type'>>): unknown[] {

```

Comment:
```suggestion
export function filterPersons(persons: Person[], personType: Person['type'], criteria: Partial<Omit<Person, 'type'>>): Person[] {

```
còn tí thì thêm vào luôn chứ tiết kiệm chi